python/cudf/cudf/core/index.py

```diff
@@ -2752,6 +2762,217 @@ def categories(self):
         return self._values.cat().categories
 
 
+def interval_range(
+    start: int = None,
+    end: int = None,
+    periods: int = None,
+    freq: int = None,
+    name: str = None,
+    closed: str = "right",
+) -> "IntervalIndex":
+    """
+    Returns a fixed frequency IntervalIndex.
+
+    Parameters
+    ----------
+    start : numeric, default None
+        Left bound for generating intervals.
+    end : numeric , default None
+        Right bound for generating intervals.
+    periods : int, default None
+        Number of periods to generate
+    freq : numeric, default None
+        The length of each interval. Must be consistent
+        with the type of start and end
+    name : str, default None
+        Name of the resulting IntervalIndex.
+    closed : {"left", "right", "both", "neither"}, default "right"
+        Whether the intervals are closed on the left-side, right-side,
+        both or neither.
+
+    Returns
+    -------
+    IntervalIndex
+
+    Examples
+    --------
+    >>> import cudf
+    >>> import pandas as pd
+    >>> cudf.interval_range(start=0,end=5)
+    IntervalIndex([(0, 0], (1, 1], (2, 2], (3, 3], (4, 4], (5, 5]],
+    ...closed='right',dtype='interval')
+    >>> cudf.interval_range(start=0,end=10, freq=2,closed='left')
+    IntervalIndex([[0, 2), [2, 4), [4, 6), [6, 8), [8, 10)],
+    ...closed='left',dtype='interval')
+    >>> cudf.interval_range(start=0,end=10, periods=3,closed='left')
+    ...IntervalIndex([[0.0, 3.3333333333333335),
+            [3.3333333333333335, 6.666666666666667),
+            [6.666666666666667, 10.0)],
+            closed='left',
+            dtype='interval')
+    """
+    if freq and periods and start and end:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and "
+            "freq, exactly three must be specified"
+        )
+    if not isinstance(start or freq or end, int) and not isinstance(
+        start or freq or end, float
+    ):
+        raise NotImplementedError("Non-numeric values not yet supported")
+    elif periods and not freq:
+        # if statement for mypy to pass
+        if end is not None and start is not None:
+            # determine if periods are float or integer
+            quotient, remainder = divmod((end - start), periods)
+            if remainder:
+                freq_step = cudf.Scalar((end - start) / periods).device_value
+            else:
+                freq_step = cudf.Scalar(quotient).device_value
+            if type(start) == freq_step.dtype:
+                start = cudf.Scalar(start).device_value
+            elif type(start) != freq_step.dtype:
+                start = cudf.Scalar(
+                    cudf.Scalar(start)._host_value.astype(freq_step.dtype)
+                ).device_value
```

Comment:
I think we should keep `freq_step` as a `cudf.Scalar` object for getting the dtype? Also we should be using `.value` instead of `._host_value`.

Reply:
i.e. something like this:
```suggestion
            if remainder:
                freq_step = cudf.Scalar((end - start) / periods)
            else:
                freq_step = cudf.Scalar(quotient)
            start = cudf.Scalar(start)
            if start.dtype != freq_step.dtype:
                start = cudf.Scalar(start.value.astype(freq_step.dtype))
            start = start.device_value
            freq_step = freq_step.device_value
```

Reply:
Ooof sorry, I'm just seeing this :/ I thought you meant I should use the overall type to typecast instead. 
Let me accept these changes instead and update :D

Reply:
Ok, added these changes in and a few more! I need to also turn on Github desktop notifications 😆